File: ProgrammersGuide/WorkingWithGmailClient/WorkingWithGmailContacts/CreateNewContact/VisualBasic/TestUser.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports System.Net


Namespace Aspose.Email.Tests.TestUtils
	Friend Enum GrantTypes
		authorization_code
		refresh_token
	End Enum
End Namespace

Namespace Aspose.Email.Tests
	Public Class TestUser
		Friend Sub New(ByVal name As String, ByVal eMail As String, ByVal password As String, ByVal domain As String)
			Me.Name = name
			Me.EMail = eMail
			Me.Password = password
			Me.Domain = domain
		End Sub

		Public ReadOnly Name As String
		Public ReadOnly EMail As String
		Public ReadOnly Password As String
		Public ReadOnly Domain As String

		Public Shared Operator =(ByVal x As TestUser, ByVal y As TestUser) As Boolean
			If CObj(x) IsNot Nothing Then
				Return x.Equals(y)
			End If
			If CObj(y) IsNot Nothing Then
				Return y.Equals(x)
			End If
			Return True
		End Operator

		Public Shared Operator <>(ByVal x As TestUser, ByVal y As TestUser) As Boolean
			Return Not(x = y)
		End Operator

		Public Shared Widening Operator CType(ByVal user As TestUser) As String
			Return If(user Is Nothing, Nothing, user.Name)
		End Operator

		Public Overrides Function GetHashCode() As Integer
			Return ToString().GetHashCode()
		End Function

		Public Overrides Overloads Function Equals(ByVal obj As Object) As Boolean
			Return obj IsNot Nothing AndAlso TypeOf obj Is TestUser AndAlso Me.ToString().Equals(obj.ToString(), StringComparison.OrdinalIgnoreCase)
		End Function

		''' <summary>
		''' Returns a string that represents the current object.
		''' </summary>
		''' <returns>A string that represents the current object.</returns>
		Public Overrides Function ToString() As String
			Return If(String.IsNullOrEmpty(Domain), Name, String.Format("{0}/{1}", Domain, Name))
		End Function
	End Class
End Namespace

File: Examples/CSharp/Gmail/TestUser.cs
// ExStart:TestUser
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Aspose.Email.Examples.CSharp.Email.Gmail
{
    internal enum GrantTypes
    {
        authorization_code,
        refresh_token
    }


    public class TestUser
    {
        internal TestUser(string name, string eMail, string password, string domain)
        {
            Name = name;
            EMail = eMail;
            Password = password;
            Domain = domain;
        }

        public readonly string Name;
        public readonly string EMail;
        public readonly string Password;
        public readonly string Domain;

        public static bool operator ==(TestUser x, TestUser y)
        {
            if ((object)x != null)
                return x.Equals(y);
            if ((object)y != null)
                return y.Equals(x);
            return true;
        }

        public static bool operator !=(TestUser x, TestUser y)
        {
            return !(x == y);
        }

        public static implicit operator string(TestUser user)
        {
            return user == null ? null : user.Name;
        }

        public override int GetHashCode()
        {
            return ToString().GetHashCode();
        }

        public override bool Equals(object obj)
        {
            return obj != null && obj is TestUser && this.ToString().Equals(obj.ToString(), StringComparison.OrdinalIgnoreCase);
        }

        public override string ToString()
        {
            return string.IsNullOrEmpty(Domain) ? Name : string.Format("{0}/{1}", Domain, Name);
        }
    }
}
// ExEnd:TestUser